modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceInfo.java

```diff
@@ -0,0 +1,170 @@
+
+    /**
+     * @param originNodeId Initiating node id.
+     * @param srvcId Service id.
+     * @param cfg Service configuration.
+     */
+    public ServiceInfo(@NotNull UUID originNodeId, @NotNull IgniteUuid srvcId, @NotNull ServiceConfiguration cfg) {
+        this(originNodeId, srvcId, cfg, false);
+    }
+
+    /**
+     * @param originNodeId Initiating node id.
+     * @param srvcId Service id.
+     * @param cfg Service configuration.
+     * @param staticCfg Statically configured flag.
+     */
+    public ServiceInfo(@NotNull UUID originNodeId, @NotNull IgniteUuid srvcId, @NotNull ServiceConfiguration cfg,
+        boolean staticCfg) {
+        this.originNodeId = originNodeId;
+        this.srvcId = srvcId;
+        this.cfg = cfg;
+        this.staticCfg = staticCfg;
+    }
+
+    /**
+     * Sets service's new topology snapshot.
+     *
+     * @param top Topology snapshot.
+     */
+    public void topologySnapshot(@NotNull Map<UUID, Integer> top) {
+        this.top = top;
+    }
+
+    /**
+     * Returns service's configuration.
+     *
+     * @return Service configuration.
+     */
+    public ServiceConfiguration configuration() {
+        return cfg;
+    }
+
+    /**
+     * @return {@code true} if statically configured.
+     */
+    public boolean staticallyConfigured() {
+        return staticCfg;
+    }
+
+    /**
+     * Rerurns services id.
+     *
+     * @return Service id.
+     */
+    public IgniteUuid serviceId() {
+        return srvcId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String name() {
+        return cfg.getName();
+    }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("unchecked")
+    @Override public Class<? extends Service> serviceClass() {
+        if (cfg instanceof LazyServiceConfiguration) {
+            String clsName = ((LazyServiceConfiguration)cfg).serviceClassName();
+
+            try {
+                return (Class<? extends Service>)Class.forName(clsName);
+            }
+            catch (ClassNotFoundException e) {
+                throw new IgniteException("Failed to find service class: " + clsName, e);
+            }
+        }
+        else
+            return cfg.getService().getClass();
+    }
+
+    /** {@inheritDoc} */
+    @Override public int totalCount() {
+        return cfg.getTotalCount();
+    }
+
+    /** {@inheritDoc} */
+    @Override public int maxPerNodeCount() {
+        return cfg.getMaxPerNodeCount();
+    }
+
+    /** {@inheritDoc} */
+    @Nullable @Override public String cacheName() {
+        return cfg.getCacheName();
+    }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("unchecked")
+    @Nullable @Override public <K> K affinityKey() {
+        return (K)cfg.getAffinityKey();
+    }
+
+    /** {@inheritDoc} */
+    @Override public UUID originNodeId() {
+        return originNodeId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Map<UUID, Integer> topologySnapshot() {
```

Comment:
Topology snapshot has a different meaning in Ignite.
If you use it here, it will overload and confuse its meaning.
I think, "service assignments" would be fine.

Reply:
`ServiceInfo` implements `ServiceDescriptor` and `#topologySnapshot` relates to the interface which I'd not changed.